Let user specify start time in config file for reproducible tests.

keep94 opened the pull request:
This PR allows tester to add a start time in the config file as seconds after 1970. When start time is present, the compliance tester always uses that start time instead of using 12 minutes before the present.  This feature enables reproducible test results.

juliusv commented on cmd/promql-compliance-tester/main.go:
```diff
@@ -96,8 +96,13 @@ func main() {
-	start := end.Add(-10 * time.Minute)
+	var start time.Time
+	if cfg.Start > 0 {
+		start = time.Unix(cfg.Start, 0)
```
Cool! Would you be up for doing parsing like here: https://github.com/prometheus/prometheus/blob/main/cmd/promtool/main.go#L693-L702

Being able to pass in also sub-second timestamps is important because otherwise everything ends up aligned at full second boundaries (since the resolution step is currently also just a multiple of seconds), potentially not catching some implementations which incorrectly always align query timestamps to seconds.

juliusv replied:
Actually, I think from my Prometheus habits perspective, it would be more intuitive to set the end time, not the start time, since that's how the Prometheus UI works as well (you pick an end time for the graph, and a range for how far back it should go).

juliusv replied:
(In the end, we should make all of `start` / `end` / `resolution` all configurable)

keep94 replied:
That all sounds good to me. I implemented everything you suggested here.

keep94 committed:
Let user specify end time, range, resolution in config file.

Signed-off-by: Travis Keep <keep94@gmail.com>

keep94 commented:
Hi, I thanks for your feedback on this PR. I have changed so that end time can be a floating point value or can be an actual time.  I also added range and resolution too.

juliusv commented on config/config.go:
```diff
@@ -14,6 +13,13 @@ type Config struct {
+	Timing                Timing        `yaml:"timing"`
+}
+
+type Timing struct {
```
I've just been thinking about whether things like this should be a flag or config value, but I guess config value makes sense for now, as all other parameters that affect test execution are also in the config file.

As for naming, how about the bit more elaborate `QueryTimeParameters` or something like that?

By the way, in the future it would be great to run instant queries (not just range queries) as well, and perhaps both types of query as part of the same compatibility test run. But we can figure that out later.

keep94 replied:
Ok I renamed Timing -> QueryTimeParameters.

keep94 committed:
Timing -> QueryTimeParameters.

Signed-off-by: Travis Keep <keep94@gmail.com>

juliusv commented:
:+1: Thanks!
